{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3582fc8",
   "metadata": {},
   "source": [
    "# CRISPR Screen Analysis\n",
    "\n",
    "This script performs the following:\n",
    "- preprocessing on fastq files\n",
    "- MAGeCK-VISPR init to create config file\n",
    "- edit config file\n",
    "- execute MAGeCK-VISPR pipeline\n",
    "\n",
    "\n",
    "Jupyter defaults to python, so each code block written in bash begins with:\n",
    "%%bash\n",
    "\n",
    "Jupyter has difficulty with conda, so each code block running MAGeCK-VISPR also begins with:\n",
    "\n",
    "source /bobross/miniconda3/etc/profile.d/conda.sh\n",
    "\n",
    "conda activate mageck-vispr\n",
    "\n",
    "cd working_directory (you have to edit this in every code block)\n",
    "\n",
    "# What is going on?\n",
    "- This script assumes you already have your fastq files and a library of guide RNAs. More about lib in the config step.\n",
    "- You will run an initialization command that will generate a config.yaml file (a file that tells MAGeCK your experimental parameters) and a Snakefile (a roadmap of conda programs to run for analysis). \n",
    "- You will edit the config file to reflect your control condition and library of guide RNAs.\n",
    "- You will test the snakefile for missing input files and then execute the analysis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9db95e29",
   "metadata": {},
   "source": [
    "# Create prelimConfig.txt and samples.txt\n",
    "This code creates 2 preliminary config files that are used to inform the rest of the jupyter script. \n",
    "\n",
    "Instructions:\n",
    "Edit your day0label, library, and list of sample names. \n",
    "\n",
    "- Library is a csv file of your grnas with their sequences and names formatted in a specific way. MAGeCK has common ones available for download on their SourceForge page (such as Brunello) or you can create your own by following the formatting. I will include the Brunello csv in the repository.\n",
    "- day0label is the fastq prefix for your negative control or presort population. If your fastq file is control.fastq, your day0label will be control. \n",
    "- samples should be listed as the prefix to fastq. So for example, if you had fastqs control.fastq, replicate1.fastq and replicate2.fastq, your samples list would be 'control', 'replicate1', 'replicate2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "49355275",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bobross/jvix/CRISPR/brunello\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "%cd /bobross/jvix/CRISPR/brunello\n",
    "day0label = 'control'\n",
    "library = 'Brunello.csv'\n",
    "samples = ['control']\n",
    "\n",
    "config = open(\"prelimConfig.txt\", \"w\")\n",
    "configelements = [day0label, library]\n",
    "for line in configelements:\n",
    "    config.write(\"%s\\n\" % line)\n",
    "config.close()\n",
    "\n",
    "samps = open(\"samples.txt\", \"w\")\n",
    "for line in samples:\n",
    "    samps.write(\"%s\\n\" % line)\n",
    "samps.close()\n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07f12145",
   "metadata": {},
   "source": [
    "# Download Fastqs\n",
    "Use this code if you don't have your fastq's yet. Edit the download command to suit your needs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4d71a941",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘fastq’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "control\n",
      "replicate1\n",
      "replicate2\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source /bobross/miniconda3/etc/profile.d/conda.sh\n",
    "cd /bobross/jvix/CRISPR/brunello\n",
    "\n",
    "FILES=$(cat samples.txt)\n",
    "mkdir fastq\n",
    "cd fastq\n",
    "\n",
    "\n",
    "for i in ${FILES}\n",
    "    do\n",
    "        #download\n",
    "        echo ${i}\n",
    "    done\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f1c6d99",
   "metadata": {},
   "source": [
    "# Preprocessing\n",
    "THIS CODE BLOCK IS MOST LIKELY NOT NECESSARY AND WILL CAUSE PROBLEMS IF YOU RUN IT. CHECK YOUR DATA FIRST.\n",
    "This code block creates reverse-complement fastq files of your pre-existing fastq files. \n",
    "\n",
    "Instructions:\n",
    "\n",
    "- Replace the file list with your fastq prefixes. For example, if you have a fastq titled C1X_R2.fastq, then you will add C1X_R2 to the list.\n",
    "- Edit the code to reflect your input and output file locations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3a4278a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "control\n",
      "replicate1\n",
      "replicate2\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source /bobross/miniconda3/etc/profile.d/conda.sh\n",
    "cd /bobross/jvix/CRISPR/brunello\n",
    "\n",
    "FILES=$(cat samples.txt)\n",
    "\n",
    "for i in ${FILES}\n",
    "    do\n",
    "        #./bin/fastx_reverse_complement -Q 33 -i scratch/FASTQ/${i}.fastq -o scratch/rcfastq/${i}.fastq\n",
    "        echo ${i}\n",
    "    done\n",
    "ls scratch/rcfastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70fcd4f4",
   "metadata": {},
   "source": [
    "# MAGeCK-VISPR init\n",
    "This code block generates a config.yaml file and displays its contents. It also generates a Snakefile.\n",
    "\n",
    "Instructions:\n",
    "- Replace the directory with your working directory\n",
    "- Replace the list of reads with your fastq files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "9b7febea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# General configuration:\n",
      "\n",
      "# Path to library design file (csv or tab-separated txt format, columns: id, sequence, gene)\n",
      "library: lib/library.csv\n",
      "# Species to use for linkouts in VISPR (e.g. mus_musculus, homo_sapiens, ...)\n",
      "species: homo_sapiens\n",
      "# Genome assembly to use for linkouts in VISPR (e.g. hg19, hg38, mm9, mm10, ...)\n",
      "assembly: hg38\n",
      "\n",
      "# Configuration of knockout target display in VISPR\n",
      "targets:\n",
      "    # if screening genes, set this to true for proper linkouts to GeneMANIA and Ensembl in VISPR\n",
      "    genes: true\n",
      "    # file with genes to hide per default in VISPR (optional, one gene per line)\n",
      "    #controls: ribosomal_genes.txt\n",
      "\n",
      "# Configuration of sgRNAs\n",
      "sgrnas:\n",
      "    # estimate sgRNA knockout efficiency during EM-procedure of MAGeCK-MLE\n",
      "    update-efficiency: false\n",
      "    # trim the 5 prime end to get rid of barcode sequences in the reads\n",
      "    # if a number (instead of AUTO) is specified, use quotes; for example:\n",
      "    # trim-5: \"0\"\n",
      "    trim-5: AUTO\n",
      "    # specify the length of the sgRNAs (without PAM sequence)\n",
      "    len: AUTO\n",
      "    # sequencing adapter that shall be removed from reads before processing with MAGeCK (optional)\n",
      "    #adapter: ACGGCTAGCTGA\n",
      "    #\n",
      "    # Use pre-computed sgrnas to annotate the library? By default it's false. \n",
      "    # Only certain assemblies (hg19, hg38, mm9, mm10) and certain sgRNA length (19, 20) are supported\n",
      "    annotate-sgrna: false\n",
      "    # Use pre-computed sgrna efficiency as an initial value of knockout efficiency?\n",
      "    # Need to set annotate-sgrna to true as well. \n",
      "    # If you need the sgRNA efficiency to be updated, set update-efficiency to true\n",
      "    annotate-sgrna-efficiency: false\n",
      "    # instead of downloading the sgrna annotation library from bitbucket,\n",
      "    # provide either the name of the file used for annotation,\n",
      "    # or the folder name where MAGeCK-VISPR will search the corresponding annotation library from that folder\n",
      "    #annotation-sgrna-file: /dev/null\n",
      "    #annotation-sgrna-folder: /src/exome_scan\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "# ATTENTION: You should and only should choose one type of input file between \"samples\" and \"counts\".\n",
      "# Configuration of samples (Cannot set \"counts\" as valid at the same time!)\n",
      "samples:\n",
      "    # The following sample information was inferred from the given FASTQ files.\n",
      "    # Adjust it according to your needs (e.g. providing descriptive sample names and grouping replicates together).\n",
      "    contorl:\n",
      "        - contorl.fastq\n",
      "\n",
      "\n",
      "# Provide paired fastq files if pair-end sequencing data is available.\n",
      "# paired:\n",
      "#     # provide a label and a paths to the paired fastq files for each sample\n",
      "#     A: path/to/A_R2.fastq\n",
      "#     B: path/to/B_R2.fastq\n",
      "#     C:\n",
      "#        - path/to/C.1_R2.fastq\n",
      "#         - path/to/C.2_R2.fastq\n",
      "\n",
      "# Specify whether report valid alignments per pair (count when both end reads mapped), when paired fastq files are provided .\n",
      "countpair: false\n",
      "\n",
      "# Instead of providing fastq files, you can also provide your own normalized (or unnormalized) count matrix (Cannot set \"samples\" as valid at the same time!).\n",
      "# If you do not want MAGeCK to normalize the counts, make sure to set up norm_methd to none\n",
      "# Support QC from count matrix\n",
      "# counts: rawcount/rawcount.txt\n",
      "\n",
      "# Provide mageck count --day0-label (optional). Multiple labels should be seperated with comma.\n",
      "day0label: plasmid\n",
      "\n",
      "\n",
      "# Provide normalization methods (none, control, median) and a list of negative control sgrnas (if norm_method=control). \n",
      "# These parameters will be applied to mageck_count, mageck_rra and mageck_mle modules.\n",
      "# Note that if this option is not specified a default median normalization will be used.\n",
      "# norm_method: control\n",
      "# control_sgrna: lib/hg19_library_1.aavs1.txt\n",
      "\n",
      "# Run mle with multi-thread. Default thread number is 1.\n",
      "# When this parameter is set (e.g., threads: 4), make sure to specify the cores in running snakemake (snakemake --cores 4)\n",
      "threads: 4\n",
      "\n",
      "# Provide a batch matrix if the samples need to be batch corrected (optional).\n",
      "# The format should be as follows (tab-separated):\n",
      "# sample          batch   covariate 1 ...\n",
      "# Sample 1        1       0           ...\n",
      "# Sample 2        2       0           ...\n",
      "# Sample 3        2       1           ...\n",
      "#\n",
      "# The first column must refer to the samples defined above.\n",
      "# batchmatrix: path/to/batchmatrix.txt\n",
      "#\n",
      "\n",
      "# copy number variation correction\n",
      "# Setting correct_cnv to true will allow CNV correction for both RRA and MLE\n",
      "#\n",
      "# to overwrite this option, you can use snakemake command line:\n",
      "# snakemake --config correct_cnv=true cnv_norm=target/cnv/file\n",
      "correct_cnv: false\n",
      "cnv_norm: /dev/null\n",
      "\n",
      "# the following cnv_cell_line is only required when performing CNV correction in RRA\n",
      "# it's not required for MLE\n",
      "cnv_cell_line: HL60\n",
      "\n",
      "\n",
      "# Additional parameters to run MLE or RRA\n",
      "# Note that additional_mle_rra_parameter is no longer used since MLE and RRA should be specified separately.\n",
      "# additional_mle_parameter: --max-sgrnapergene-permutation 30\n",
      "# additional_rra_parameter: --remove-zero both --additional-rra-parameters \"--max-sgrnapergene-permutation 200\"\n",
      "\n",
      "# Configuration of experiments.\n",
      "# An experiment defines the comparison that shall be analyzed with MAGeCK.\n",
      "# You can define as many experiments as you want.\n",
      "# You can define either MAGeCK-RRA or MAGeCK-MLE experiments, but cannot define both in a single configuration file.\n",
      "experiments:\n",
      "    # provide a descriptive name for your experiment (it will show up in VISPR)\n",
      "    \"mle\":\n",
      "        # This is a MAGeCK-MLE experiment.\n",
      "        # Here, users can either specify a day0label, or a design matrix file (see http://mageck.sourceforge.net for details).\n",
      "        # if day0label is specified, use an empty file (like /dev/null) as a design matrix file.\n",
      "        # Sample names in the design matrix must refer to the samples defined above.\n",
      "        designmatrix: /dev/null\n",
      "    #\"rra\":\n",
      "        # This is a MAGeCK-RRA experiment.\n",
      "        # You must specify treatment and control samples.\n",
      "        # The sample names must refer to the samples defined above in the\n",
      "        # samples section.\n",
      "        #treatment:\n",
      "        #    - A\n",
      "        #control:\n",
      "        #    - B\n",
      "        #    - C"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source /bobross/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate mageck-vispr\n",
    "cd /bobross/jvix/CRISPR/brunello\n",
    "\n",
    "# mageck-vispr init /path/to/working/directory --reads file/path/fastq1 file/path/fastq2 ...\n",
    "mageck-vispr init /bobross/jvix/CRISPR/brunello --reads contorl.fastq  \n",
    "\n",
    "\n",
    "#view config file\n",
    "cat config.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "091d22c0",
   "metadata": {},
   "source": [
    "# Edit config file\n",
    "This code block uses python and preliminConfig.txt to edit your config file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1bc3e6d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "# General configuration:\n",
      "# Path to library design file (csv or tab-separated txt format, columns: id, sequence, gene)\n",
      "Brunello.csv\n",
      "# Species to use for linkouts in VISPR (e.g. mus_musculus, homo_sapiens, ...)\n",
      "species: homo_sapiens\n",
      "# Genome assembly to use for linkouts in VISPR (e.g. hg19, hg38, mm9, mm10, ...)\n",
      "assembly: hg38\n",
      "# Configuration of knockout target display in VISPR\n",
      "targets:\n",
      "    # if screening genes, set this to true for proper linkouts to GeneMANIA and Ensembl in VISPR\n",
      "    genes: true\n",
      "    # file with genes to hide per default in VISPR (optional, one gene per line)\n",
      "    #controls: ribosomal_genes.txt\n",
      "# Configuration of sgRNAs\n",
      "sgrnas:\n",
      "    # estimate sgRNA knockout efficiency during EM-procedure of MAGeCK-MLE\n",
      "    update-efficiency: false\n",
      "    # trim the 5 prime end to get rid of barcode sequences in the reads\n",
      "    # if a number (instead of AUTO) is specified, use quotes; for example:\n",
      "    # trim-5: \"0\"\n",
      "    trim-5: AUTO\n",
      "    # specify the length of the sgRNAs (without PAM sequence)\n",
      "    len: AUTO\n",
      "    # sequencing adapter that shall be removed from reads before processing with MAGeCK (optional)\n",
      "    #adapter: ACGGCTAGCTGA\n",
      "    #\n",
      "    # Use pre-computed sgrnas to annotate the library? By default it's false. \n",
      "    # Only certain assemblies (hg19, hg38, mm9, mm10) and certain sgRNA length (19, 20) are supported\n",
      "    annotate-sgrna: false\n",
      "    # Use pre-computed sgrna efficiency as an initial value of knockout efficiency?\n",
      "    # Need to set annotate-sgrna to true as well. \n",
      "    # If you need the sgRNA efficiency to be updated, set update-efficiency to true\n",
      "    annotate-sgrna-efficiency: false\n",
      "    # instead of downloading the sgrna annotation library from bitbucket,\n",
      "    # provide either the name of the file used for annotation,\n",
      "    # or the folder name where MAGeCK-VISPR will search the corresponding annotation library from that folder\n",
      "    #annotation-sgrna-file: /dev/null\n",
      "    #annotation-sgrna-folder: /src/exome_scan\n",
      "# ATTENTION: You should and only should choose one type of input file between \"samples\" and \"counts\".\n",
      "# Configuration of samples (Cannot set \"counts\" as valid at the same time!)\n",
      "samples:\n",
      "    # The following sample information was inferred from the given FASTQ files.\n",
      "    # Adjust it according to your needs (e.g. providing descriptive sample names and grouping replicates together).\n",
      "    contorl:\n",
      "        - contorl.fastq\n",
      "# Provide paired fastq files if pair-end sequencing data is available.\n",
      "# paired:\n",
      "#     # provide a label and a paths to the paired fastq files for each sample\n",
      "#     A: path/to/A_R2.fastq\n",
      "#     B: path/to/B_R2.fastq\n",
      "#     C:\n",
      "#        - path/to/C.1_R2.fastq\n",
      "#         - path/to/C.2_R2.fastq\n",
      "# Specify whether report valid alignments per pair (count when both end reads mapped), when paired fastq files are provided .\n",
      "countpair: false\n",
      "# Instead of providing fastq files, you can also provide your own normalized (or unnormalized) count matrix (Cannot set \"samples\" as valid at the same time!).\n",
      "# If you do not want MAGeCK to normalize the counts, make sure to set up norm_methd to none\n",
      "# Support QC from count matrix\n",
      "# counts: rawcount/rawcount.txt\n",
      "# Provide mageck count --day0-label (optional). Multiple labels should be seperated with comma.\n",
      "day0label:control\n",
      "# Provide normalization methods (none, control, median) and a list of negative control sgrnas (if norm_method=control). \n",
      "# These parameters will be applied to mageck_count, mageck_rra and mageck_mle modules.\n",
      "# Note that if this option is not specified a default median normalization will be used.\n",
      "# norm_method: control\n",
      "# control_sgrna: lib/hg19_library_1.aavs1.txt\n",
      "# Run mle with multi-thread. Default thread number is 1.\n",
      "# When this parameter is set (e.g., threads: 4), make sure to specify the cores in running snakemake (snakemake --cores 4)\n",
      "threads: 4\n",
      "# Provide a batch matrix if the samples need to be batch corrected (optional).\n",
      "# The format should be as follows (tab-separated):\n",
      "# sample          batch   covariate 1 ...\n",
      "# Sample 1        1       0           ...\n",
      "# Sample 2        2       0           ...\n",
      "# Sample 3        2       1           ...\n",
      "#\n",
      "# The first column must refer to the samples defined above.\n",
      "# batchmatrix: path/to/batchmatrix.txt\n",
      "#\n",
      "# copy number variation correction\n",
      "# Setting correct_cnv to true will allow CNV correction for both RRA and MLE\n",
      "#\n",
      "# to overwrite this option, you can use snakemake command line:\n",
      "# snakemake --config correct_cnv=true cnv_norm=target/cnv/file\n",
      "correct_cnv: false\n",
      "cnv_norm: /dev/null\n",
      "# the following cnv_cell_line is only required when performing CNV correction in RRA\n",
      "# it's not required for MLE\n",
      "cnv_cell_line: HL60\n",
      "# Additional parameters to run MLE or RRA\n",
      "# Note that additional_mle_rra_parameter is no longer used since MLE and RRA should be specified separately.\n",
      "# additional_mle_parameter: --max-sgrnapergene-permutation 30\n",
      "# additional_rra_parameter: --remove-zero both --additional-rra-parameters \"--max-sgrnapergene-permutation 200\"\n",
      "# Configuration of experiments.\n",
      "# An experiment defines the comparison that shall be analyzed with MAGeCK.\n",
      "# You can define as many experiments as you want.\n",
      "# You can define either MAGeCK-RRA or MAGeCK-MLE experiments, but cannot define both in a single configuration file.\n",
      "experiments:\n",
      "    # provide a descriptive name for your experiment (it will show up in VISPR)\n",
      "    \"mle\":\n",
      "        # This is a MAGeCK-MLE experiment.\n",
      "        # Here, users can either specify a day0label, or a design matrix file (see http://mageck.sourceforge.net for details).\n",
      "        # if day0label is specified, use an empty file (like /dev/null) as a design matrix file.\n",
      "        # Sample names in the design matrix must refer to the samples defined above.\n",
      "        designmatrix: /dev/null\n",
      "    #\"rra\":\n",
      "        # This is a MAGeCK-RRA experiment.\n",
      "        # You must specify treatment and control samples.\n",
      "        # The sample names must refer to the samples defined above in the\n",
      "        # samples section.\n",
      "        #treatment:\n",
      "        #    - A\n",
      "        #control:\n",
      "        #    - B\n",
      "        #    - C\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "config = pd.read_table('config.yaml', header=None, delimiter=None)[0].to_list()\n",
    "day0lib = pd.read_table('prelimConfig.txt', header=None, delimiter=None)[0].to_list()\n",
    "samples = pd.read_table('samples.txt', header=None, delimiter=None)[0].to_list()\n",
    "\n",
    "#print(day0lib)\n",
    "#print(samples)\n",
    "config[2] = day0lib[1]\n",
    "config[58+2*(len(samples))] = 'day0label:'+ day0lib[0]\n",
    "print(len(samples))\n",
    "os.remove('config.yaml')\n",
    "\n",
    "fp = open('config.yaml', 'w')\n",
    "for line in config:\n",
    "    print(line)\n",
    "    fp.write(\"%s\\n\" % line)\n",
    "fp.close()\n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2655228e",
   "metadata": {},
   "source": [
    "# Perform a dry run followed by analysis\n",
    "Instructions:\n",
    "- Change the directory to your working directory\n",
    "- Make sure the first code block doesn't get angry with you about nonexistent files or other errors.\n",
    "- Run the second code block. This could take up to a few hours. It usually hangs out at ~62% depending on your sample number for a million years. Don't be concerned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4123182c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building DAG of jobs...\n",
      "Job stats:\n",
      "job             count    min threads    max threads\n",
      "------------  -------  -------------  -------------\n",
      "all                 1              1              1\n",
      "fastqc              3              1              1\n",
      "mageck_count        1              1              1\n",
      "mageck_mle          1              1              1\n",
      "mageck_rra          1              1              1\n",
      "vispr               1              1              1\n",
      "total               8              1              1\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:08:51 2023]\n",
      "rule fastqc:\n",
      "    input: rcfastq/S1H_R2.fastq\n",
      "    output: results/qc/S1H_R2_0\n",
      "    log: logs/fastqc/S1H_R2_0.log\n",
      "    jobid: 6\n",
      "    reason: Updated input files: rcfastq/S1H_R2.fastq\n",
      "    wildcards: replicate=S1H_R2_0\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:08:51 2023]\n",
      "rule fastqc:\n",
      "    input: rcfastq/C1X_R2.fastq\n",
      "    output: results/qc/C1X_R2_0\n",
      "    log: logs/fastqc/C1X_R2_0.log\n",
      "    jobid: 5\n",
      "    reason: Updated input files: rcfastq/C1X_R2.fastq\n",
      "    wildcards: replicate=C1X_R2_0\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:08:51 2023]\n",
      "rule mageck_count:\n",
      "    input: rcfastq/C1X_R2.fastq, rcfastq/S1H_R2.fastq, rcfastq/S2L_R2.fastq, sabatini.csv\n",
      "    output: results/count/all.count.txt, results/count/all.count_normalized.txt, results/count/all.countsummary.txt\n",
      "    log: logs/mageck/count/all.log\n",
      "    jobid: 3\n",
      "    reason: Missing output files: results/count/all.count.txt, results/count/all.countsummary.txt, results/count/all.count_normalized.txt\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:08:51 2023]\n",
      "rule fastqc:\n",
      "    input: rcfastq/S2L_R2.fastq\n",
      "    output: results/qc/S2L_R2_0\n",
      "    log: logs/fastqc/S2L_R2_0.log\n",
      "    jobid: 7\n",
      "    reason: Updated input files: rcfastq/S2L_R2.fastq\n",
      "    wildcards: replicate=S2L_R2_0\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:08:51 2023]\n",
      "rule mageck_mle:\n",
      "    input: results/count/all.count.txt, /dev/null\n",
      "    output: results/test/mle.gene_summary.txt, results/test/mle.sgrna_summary.txt\n",
      "    log: logs/mageck/test/mle.log\n",
      "    jobid: 2\n",
      "    reason: Missing output files: results/test/mle.gene_summary.txt, results/test/mle.sgrna_summary.txt; Input files updated by another job: results/count/all.count.txt\n",
      "    wildcards: experiment=mle\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:08:51 2023]\n",
      "rule mageck_rra:\n",
      "    input: results/count/all.count.txt\n",
      "    output: results/test/mle.rra.gene_summary.txt, results/test/mle.rra.sgrna_summary.txt, results/test/mle.rra.S1H_R2_vs_C1X_R2.sgrna_summary.txt, results/test/mle.rra.S2L_R2_vs_C1X_R2.sgrna_summary.txt\n",
      "    log: logs/mageck/test/mle.rra.log\n",
      "    jobid: 4\n",
      "    reason: Missing output files: results/test/mle.rra.gene_summary.txt; Input files updated by another job: results/count/all.count.txt\n",
      "    wildcards: experiment=mle.rra\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:08:51 2023]\n",
      "rule vispr:\n",
      "    input: results/test/mle.gene_summary.txt, results/test/mle.rra.gene_summary.txt, results/test/mle.sgrna_summary.txt, results/count/all.count_normalized.txt, results/count/all.countsummary.txt, results/qc/C1X_R2_0, results/qc/S1H_R2_0, results/qc/S2L_R2_0\n",
      "    output: results/mle.vispr.yaml\n",
      "    jobid: 1\n",
      "    reason: Missing output files: results/mle.vispr.yaml; Input files updated by another job: results/qc/S2L_R2_0, results/count/all.count_normalized.txt, results/test/mle.sgrna_summary.txt, results/qc/C1X_R2_0, results/count/all.countsummary.txt, results/test/mle.gene_summary.txt, results/test/mle.rra.gene_summary.txt, results/qc/S1H_R2_0\n",
      "    wildcards: experiment=mle\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "[Fri Jul  7 09:08:51 2023]\n",
      "localrule all:\n",
      "    input: results/mle.vispr.yaml\n",
      "    jobid: 0\n",
      "    reason: Input files updated by another job: results/mle.vispr.yaml\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "Job stats:\n",
      "job             count    min threads    max threads\n",
      "------------  -------  -------------  -------------\n",
      "all                 1              1              1\n",
      "fastqc              3              1              1\n",
      "mageck_count        1              1              1\n",
      "mageck_mle          1              1              1\n",
      "mageck_rra          1              1              1\n",
      "vispr               1              1              1\n",
      "total               8              1              1\n",
      "\n",
      "Reasons:\n",
      "    (check individual jobs above for details)\n",
      "    input files updated by another job:\n",
      "        all, mageck_mle, mageck_rra, vispr\n",
      "    missing output files:\n",
      "        mageck_count, mageck_mle, mageck_rra, vispr\n",
      "    updated input files:\n",
      "        fastqc\n",
      "\n",
      "This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source /bobross/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate mageck-vispr\n",
    "\n",
    "\n",
    "\n",
    "snakemake -n\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "04058fce",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building DAG of jobs...\n",
      "Using shell: /bin/bash\n",
      "Provided cores: 8\n",
      "Rules claiming more threads will be scaled down.\n",
      "Job stats:\n",
      "job             count    min threads    max threads\n",
      "------------  -------  -------------  -------------\n",
      "all                 1              1              1\n",
      "fastqc              3              1              1\n",
      "mageck_count        1              1              1\n",
      "mageck_mle          1              1              1\n",
      "mageck_rra          1              1              1\n",
      "vispr               1              1              1\n",
      "total               8              1              1\n",
      "\n",
      "Select jobs to execute...\n",
      "\n",
      "[Fri Jul  7 09:09:02 2023]\n",
      "rule fastqc:\n",
      "    input: rcfastq/C1X_R2.fastq\n",
      "    output: results/qc/C1X_R2_0\n",
      "    log: logs/fastqc/C1X_R2_0.log\n",
      "    jobid: 5\n",
      "    reason: Updated input files: rcfastq/C1X_R2.fastq\n",
      "    wildcards: replicate=C1X_R2_0\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:09:02 2023]\n",
      "rule fastqc:\n",
      "    input: rcfastq/S1H_R2.fastq\n",
      "    output: results/qc/S1H_R2_0\n",
      "    log: logs/fastqc/S1H_R2_0.log\n",
      "    jobid: 6\n",
      "    reason: Updated input files: rcfastq/S1H_R2.fastq\n",
      "    wildcards: replicate=S1H_R2_0\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:09:02 2023]\n",
      "rule mageck_count:\n",
      "    input: rcfastq/C1X_R2.fastq, rcfastq/S1H_R2.fastq, rcfastq/S2L_R2.fastq, sabatini.csv\n",
      "    output: results/count/all.count.txt, results/count/all.count_normalized.txt, results/count/all.countsummary.txt\n",
      "    log: logs/mageck/count/all.log\n",
      "    jobid: 3\n",
      "    reason: Missing output files: results/count/all.count.txt, results/count/all.countsummary.txt, results/count/all.count_normalized.txt\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:09:02 2023]\n",
      "rule fastqc:\n",
      "    input: rcfastq/S2L_R2.fastq\n",
      "    output: results/qc/S2L_R2_0\n",
      "    log: logs/fastqc/S2L_R2_0.log\n",
      "    jobid: 7\n",
      "    reason: Updated input files: rcfastq/S2L_R2.fastq\n",
      "    wildcards: replicate=S2L_R2_0\n",
      "    resources: tmpdir=/tmp\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for S2L_R2.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Fri Jul  7 09:09:54 2023]\n",
      "Finished job 7.\n",
      "1 of 8 steps (12%) done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for C1X_R2.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Fri Jul  7 09:25:45 2023]\n",
      "Finished job 5.\n",
      "2 of 8 steps (25%) done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for S1H_R2.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Fri Jul  7 09:26:04 2023]\n",
      "Finished job 6.\n",
      "3 of 8 steps (38%) done\n",
      "[Fri Jul  7 09:28:58 2023]\n",
      "Finished job 3.\n",
      "4 of 8 steps (50%) done\n",
      "Select jobs to execute...\n",
      "\n",
      "[Fri Jul  7 09:28:59 2023]\n",
      "rule mageck_mle:\n",
      "    input: results/count/all.count.txt, /dev/null\n",
      "    output: results/test/mle.gene_summary.txt, results/test/mle.sgrna_summary.txt\n",
      "    log: logs/mageck/test/mle.log\n",
      "    jobid: 2\n",
      "    reason: Missing output files: results/test/mle.gene_summary.txt, results/test/mle.sgrna_summary.txt; Input files updated by another job: results/count/all.count.txt\n",
      "    wildcards: experiment=mle\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "\n",
      "[Fri Jul  7 09:28:59 2023]\n",
      "rule mageck_rra:\n",
      "    input: results/count/all.count.txt\n",
      "    output: results/test/mle.rra.gene_summary.txt, results/test/mle.rra.sgrna_summary.txt, results/test/mle.rra.S1H_R2_vs_C1X_R2.sgrna_summary.txt, results/test/mle.rra.S2L_R2_vs_C1X_R2.sgrna_summary.txt\n",
      "    log: logs/mageck/test/mle.rra.log\n",
      "    jobid: 4\n",
      "    reason: Missing output files: results/test/mle.rra.gene_summary.txt; Input files updated by another job: results/count/all.count.txt\n",
      "    wildcards: experiment=mle.rra\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "[Fri Jul  7 09:29:42 2023]\n",
      "Finished job 4.\n",
      "5 of 8 steps (62%) done\n",
      "[Fri Jul  7 12:35:49 2023]\n",
      "Finished job 2.\n",
      "6 of 8 steps (75%) done\n",
      "Select jobs to execute...\n",
      "\n",
      "[Fri Jul  7 12:35:49 2023]\n",
      "rule vispr:\n",
      "    input: results/test/mle.gene_summary.txt, results/test/mle.rra.gene_summary.txt, results/test/mle.sgrna_summary.txt, results/count/all.count_normalized.txt, results/count/all.countsummary.txt, results/qc/C1X_R2_0, results/qc/S1H_R2_0, results/qc/S2L_R2_0\n",
      "    output: results/mle.vispr.yaml\n",
      "    jobid: 1\n",
      "    reason: Missing output files: results/mle.vispr.yaml; Input files updated by another job: results/qc/S2L_R2_0, results/test/mle.gene_summary.txt, results/test/mle.rra.gene_summary.txt, results/count/all.countsummary.txt, results/qc/C1X_R2_0, results/qc/S1H_R2_0, results/count/all.count_normalized.txt, results/test/mle.sgrna_summary.txt\n",
      "    wildcards: experiment=mle\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "[Fri Jul  7 12:35:54 2023]\n",
      "Finished job 1.\n",
      "7 of 8 steps (88%) done\n",
      "Select jobs to execute...\n",
      "\n",
      "[Fri Jul  7 12:35:54 2023]\n",
      "localrule all:\n",
      "    input: results/mle.vispr.yaml\n",
      "    jobid: 0\n",
      "    reason: Input files updated by another job: results/mle.vispr.yaml\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "[Fri Jul  7 12:35:54 2023]\n",
      "Finished job 0.\n",
      "8 of 8 steps (100%) done\n",
      "Complete log: .snakemake/log/2023-07-07T090901.123051.snakemake.log\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source /bobross/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate mageck-vispr\n",
    "snakemake --cores 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccd30fd9",
   "metadata": {},
   "source": [
    "#  Run VISPR\n",
    "You MUST manually stop the code block when you are done!!! Closing jupyter without stopping the code block will leave the process running in the background on the server."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3751acdb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Starting server.\n",
      "\n",
      "Open:  go to 132.198.80.178:8079 in your browser.\n",
      "Note: Safari and Internet Explorer are currently unsupported.\n",
      "Close: hit Ctrl-C in this terminal.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading data.\n",
      " * Serving Flask app 'vispr.server'\n",
      " * Debug mode: off\n",
      "Error while terminating subprocess (pid=3358220): \n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source /bobross/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate mageck-vispr\n",
    "nohup vispr server results/mle.vispr.yaml --host 132.198.80.178 --port 8079"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a48c61d8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
